interfaces/apparmor/spec_test.go

```diff
@@ -280,3 +280,75 @@ func (s *specSuite) TestApparmorSnippetsFromLayout(c *C) {
 	c.Assert(updateNS[3], Equals, profile3)
 	c.Assert(updateNS, DeepEquals, []string{profile0, profile1, profile2, profile3})
 }
+
+func (s *specSuite) TestChopTree(c *C) {
+	for _, tc := range []struct {
+		p    string   // path
+		d    int      // depth
+		l, r []string // left and right path expressions
+		e    string   // error pattern, if non-empty
+	}{
+		// Test case from the documentation of the function.
+		{p: "/foo/bar/froz/baz/", d: 3, // Assume first three directories exist
+			l: []string{"/", "/foo/", "/foo/bar/"},
+			// Assume that /foo/bar/froz and beyond may be missing
+			r: []string{"/foo/bar/*", "/foo/bar/*/", "/foo/bar/froz/*", "/foo/bar/froz/*/"}},
+
+		// Exhaustive test cases for directory paths.
+
+		{p: "/foo/bar/froz/", d: 0, // Assume that no directories exist
+			r: []string{"/*", "/*/", "/foo/*", "/foo/*/", "/foo/bar/*", "/foo/bar/*/"}},
+		{p: "/foo/bar/froz/", d: 1, // Assume that the root directory exists
+			l: []string{"/"},
+			r: []string{"/*", "/*/", "/foo/*", "/foo/*/", "/foo/bar/*", "/foo/bar/*/"}},
+		{p: "/foo/bar/froz/", d: 2, // Assume that /foo/ exists.
+			l: []string{"/", "/foo/"},
+			r: []string{"/foo/*", "/foo/*/", "/foo/bar/*", "/foo/bar/*/"}},
+		{p: "/foo/bar/froz/", d: 3, // Assume that /foo/bar/ exists.
+			l: []string{"/", "/foo/", "/foo/bar/"},
+			r: []string{"/foo/bar/*", "/foo/bar/*/"}},
+		{p: "/foo/bar/froz/", d: 4, // Assume that /foo/bar/froz/ exists.
+			l: []string{"/", "/foo/", "/foo/bar/", "/foo/bar/froz/"}},
+
+		// Exhaustive test cases for file paths.
+
+		{p: "/foo/bar/froz", d: 0, // Assume that no directories exist
+			l: []string(nil),
+			r: []string{"/*", "/*/", "/foo/*", "/foo/*/", "/foo/bar/*", "/foo/bar/*/"}},
+		{p: "/foo/bar/froz", d: 1, // Assume that the root directory exists
+			l: []string{"/"},
+			r: []string{"/*", "/*/", "/foo/*", "/foo/*/", "/foo/bar/*", "/foo/bar/*/"}},
+		{p: "/foo/bar/froz", d: 2, // Assume that /foo/ exists.
+			l: []string{"/", "/foo/"},
+			r: []string{"/foo/*", "/foo/*/", "/foo/bar/*", "/foo/bar/*/"}},
+		{p: "/foo/bar/froz", d: 3, // Assume that /foo/bar/ exists.
+			l: []string{"/", "/foo/", "/foo/bar/"},
+			r: []string{"/foo/bar/*", "/foo/bar/*/"}},
+		{p: "/foo/bar/froz", d: 4, // Assume that /foo/bar/froz exists.
+			l: []string{"/", "/foo/", "/foo/bar/", "/foo/bar/froz"}},
+
+		// Assumed prefix depth larger than actual path depth is harmless.
+		{p: "/foo/bar/froz/", d: 5,
+			l: []string{"/", "/foo/", "/foo/bar/", "/foo/bar/froz/"}},
+		{p: "/foo/bar/froz", d: 5,
+			l: []string{"/", "/foo/", "/foo/bar/", "/foo/bar/froz"}},
+
+		// Unclean paths are not allowed.
+		{p: "/foo/../bar", d: 1, e: "cannot chop unclean path: .*"},
+		{p: "/foo//bar", d: 1, e: "cannot chop unclean path: .*"},
```

Comment:
Add relative check `{p: "foo/bar", d: 1, ...`?

Reply:
This is done, the comment shows because I added it below.